### DISCUSSION TITLE
How to transform global axis into local using python in cadwork?

### DISCUSSION THREAD

**Discussion by lucas-ita** [cwapi3d/cwapi3dpython#129] [General] [answered]
Hi,
Could someone please share a code snippet utilizing the `apply_transformation_coordinate(elements, old_point, old_xl, old_yl, new_point, new_xl, new_yl)` method? To aid our understanding, let's consider a practical example involving an inclined beam in Cadwork. Let's envision placing the beam in a scene and, subsequently, using the mentioned method to transform the global axis into local coordinates. The expected outcome is that the beam no longer remains inclined.

**Comment by Brunner246** [comment, marked as answer]
you could do something like this

```python
from typing import List

import cadwork
import element_controller as ec
import geometry_controller as gc


def get_element_orientation_and_position(element_id: int):
    point1 = gc.get_p1(element_id)
    x_direction = gc.get_xl(element_id)
    y_direction = gc.get_yl(element_id)
    return point1, x_direction, y_direction


def transform_element(element_id: int, point1_new: cadwork.point_3d, x_direction_new: cadwork.point_3d,
                      y_direction_new: cadwork.point_3d):
    point1_old, x_direction_old, y_direction_old = get_element_orientation_and_position(element_id)
    ec.apply_transformation_coordinate([element_id],
                                       point1_old,
                                       x_direction_old,
                                       y_direction_old,
                                       point1_new,
                                       x_direction_new,
                                       y_direction_new)


if __name__ == '__main__':
    element_ids: List[int] = ec.get_user_element_ids()

    x_direction_new = cadwork.point_3d(1., 0., 0.)
    y_direction_new = cadwork.point_3d(0., 1., 0.)

    for element_id in element_ids:
        point1_new = gc.get_p1(element_id) # or set your preferred position
        transform_element(element_id, point1_new, x_direction_new, y_direction_new)

```

**Reply by roehligas** [reply]
Has Cadwork considered expanding the point_3d functionality to have standard solutions for the most fundamental vector problems?

It would make a lot of my scripts more readable as I'm always rewriting the same functionality on the run pretty much...

**Reply by Brunner246** [reply]
Can you specify what is meant with `the most fundamental vector problems` ?

https://docs.cadwork.com/projects/cwapi3dpython/en/latest/examples/geometry/

There is also https://github.com/cwapi3d/cwmath which is meant to be an open source library from and for users

**Reply by roehligas** [reply]
A function I haven't been able to write yet is to be able to get rotation axis and angle based on 2 inputvectors in one go.
So far I'm always doing it axis per axis because it was faster to write. I'm not gonna share it because it's nothing that should be replicated the way i did it (not even sure why it works...) But the functionality comes in super handy when you want to align existing elements according to a given vector or do some verifications of alignment...

Apart from that, those are the functions frequently rewrite or copy/paste. :)
```
def angle_between_vectors(v1:cw.point_3d, v2:cw.point_3d) -> float:
    if (v1.magnitude() * v2.magnitude()) == 0:
        return math.pi/2
    return math.acos(v1.dot(v2) / (v1.magnitude() * v2.magnitude()))
```

```
def project_point_on_line(linepoint1, linepoint2, point):
    """
        Projects a 3d point onto a 3d line.
        Returns the closest point on the line.
    """
    line_vector = linepoint2 - linepoint1

    point_vector = point - linepoint1

    t = point_vector.dot(line_vector) / line_vector.dot(line_vector)

    projected_point = linepoint1 + t * line_vector

    return projected_point
```

```
def project_point_on_plane(point, plane_normal, plane_point):
    x, y, z = point
    a, b, c = plane_normal
    p, q, r = plane_point

    # Calculate the perpendicular distance from the point to the plane
    distance = (a * (x - p) + b * (y - q) + c * (z - r)) / (a**2 + b**2 + c**2)

    # Calculate the projected point

    x_proj = x - distance * a
    y_proj = y - distance * b
    z_proj = z - distance * c

    projected_point = cadwork.point_3d(x_proj, y_proj, z_proj)
    
    return projected_point

```

**Reply by Brunner246** [reply]
I havent tested it so far, but maybe that could be something for you https://github.com/sympy/sympy

**Reply by roehligas** [reply]
I honestly haven't tried sympy within cadwork yet, I assumed it would be rather slow. I use it a lot outside of cadwork though. I'll look into this!

But I still think that the above mentioned functionality should be a part of the cadwork standard library, it would make the code so much cleaner for me personally... :)

**Reply by Brunner246** [reply]
we want to keept the `cwapi3d` consistent with just cadwork functions. But we can start with a vector class in `cwmath`.

**Reply by Brunner246** [reply]
Unfortunately, the `cwmath` library has not undergone much further development. I have now added a few initial examples. We will merge this soon. As I said, the goal is for this to be developed further with the community ;)

https://github.com/cwapi3d/cwmath/tree/feature/vector-plane-class/src/cwmath

**Reply by roehligas** [reply]
That's the spirit!

**Comment by lucas-ita** [comment]
I have a question. Let's imagine the following scenario: we have a dictionary called element_collision_and_contact, where each key represents a timber element in Cadwork (timber_element), and the values are the elements that collide or contact with this timber element (conFilter). I already understand how to transform timber_element from global to local coordinates, but I'm having difficulty maintaining the relative position of connFilter with respect to this element. How can I perform this transformation properly?
For the sake of having an accompanying intuition: `element_collision_and_contact[timber_element] = connFilter.

**Reply by roehligas** [reply]
I don't understand the question.

1. You have managed to find colliding elements and store this in a dictionary right? Or is this the problem how to do it?
2. Why would those collisions be relevant in a local coordinate system? Wouldn't it make more sense to calculate/store them globally and recalculate as local on demand?
3. Do you have some images that visualize what you're doing? It's hard to follow your thought process.
4. If you post some code it's easier to help. :)

**Reply by lucas-ita** [reply]
Of course
at the moment:
![image](https://github.com/cwapi3d/cwapi3dpython/assets/123772761/d1ce8a46-6b9d-4372-bee6-0b657e98820a)
how I would like it to be:
![image](https://github.com/cwapi3d/cwapi3dpython/assets/123772761/9cc6c32f-32ff-41da-8bd0-163cbe9addd3)
I can transform timber_element from global to local, the problem is maintaining the relative position of the elements that touch/collide it (which is in red in the image).

I have already established in the dictionary below the proper correspondence between timber_element and the elements that touch/collide with it.

`element_collision_and_contact[timber_element] = conFilter`

I will provide a quick example (using what has already been sent on this page) to transform timber_element from global coordinates to local coordinates, based on the dictionary above. I still need to include the logic to ensure that the elements that touch/collide with it (conFilter) follow the relative position of the piece.

```
def get_element_orientation_and_position(element_id: int):
    point1 = gc.get_p1(element_id)
    x_direction = gc.get_xl(element_id)
    y_direction = gc.get_yl(element_id)
    return point1, x_direction, y_direction


def transform_element(element_id: int, point1_new: cadwork.point_3d, x_direction_new: cadwork.point_3d,
                      y_direction_new: cadwork.point_3d):
    point1_old, x_direction_old, y_direction_old = get_element_orientation_and_position(element_id)
    ec.apply_transformation_coordinate([element_id],
                                       point1_old,
                                       x_direction_old,
                                       y_direction_old,
                                       point1_new,
                                       x_direction_new,
                                       y_direction_new)


x_direction_new = cadwork.point_3d(1., 0., 0.)
y_direction_new = cadwork.point_3d(0., 1., 0.)

for timber_element, con_elements in element_collision_and_contact.items():
point1_new = gc.get_p1(timber_element)
transform_element(timber_element, point1_new, x_direction_new, y_direction_new)
```
How can we ensure that the elements that touch/collide with it maintain their relative position to the piece?

**Reply by roehligas** [reply]
_How can we ensure that the elements that touch/collide with it maintain their relative position to the piece?_

I'm sorry I don't understand your goals.
If everything shall be rotate, you should rotate everything?! This means sending in all element_ids instead of just a single element?

**Reply by lucas-ita** [reply]
Let's say you have a sloped beam in Cadwork, and this beam has bolts and plates along its length.

In other words, what I'm trying to do is rotate the beam and those elements that collide and touch (bolts and plates) so that the beam is no longer sloped, and the bolts and plates along it move accordingly.

The idea of transitioning from global to local coordinates is a way to achieve this. However, it becomes a bit challenging to ensure that the bolts and plates follow the transformation of the beam.

**Reply by roehligas** [reply]
Do you want to just do this or do other things with the script after?

**Reply by lucas-ita** [reply]
Figuring out a way to do this in an efficient way would be enough for me

**Reply by roehligas** [reply]
The script below is a script we used for some truck stacking a year ago, rather simple but it should do the job if i understand correctly.

Make sure the elements that belong together share the same subgroup for testing. Then mark those elements and run the script.
It will align those elements in XY-plane by the coordinates of the biggest element.


```
"""
Script for Creating Export Solid based on Subgroup values

Written by RR 24.02.2023

Changelog:
23-02-24:       - Grouping functionality, sorting elements to determine main object by biggest volume.
23-03-03:       - Added functionality:
                    => Copies name and all other attributes from the main object
23-03-07:       - Added functionality:
                    => Distinguishes between columns and beams
"""



# import modules
import  cadwork                 as cw
import  element_controller      as ec
import  attribute_controller    as ac
import geometry_controller      as gc
import menu_controller          as mec 
import utility_controller       as uc
import visualization_controller as vc
import math

splitted_glulam_endings = ['_A','_B','_C','_D','_E','_F']

MAX_NUM_USER_ATTRIBUTES = 30

EXPORT_SOLID_STANDARD_NAME_COLUMN = 'Export Glulam_Column'
EXPORT_SOLID_STANDARD_NAME_BEAM = 'Export Glulam_Beam'

def group_elements_by_function(elements_in,function_in):
    elements_dict = {}

    for curr_id in elements_in:
        curr_attribute = function_in(curr_id)
        
        if curr_attribute not in elements_dict:
            elements_dict[curr_attribute] = []
        
        elements_dict[curr_attribute].append(curr_id)

    return elements_dict


def sort_elements_by_function(elements_in, function_in):
    sorted_elements =  elements_in
    
    sorted_elements.sort(key=function_in)
    
    return sorted_elements





def angle_between_vectors(v1:cw.point_3d, v2:cw.point_3d) -> float:
    return math.acos(v1.dot(v2) / (v1.magnitude() * v2.magnitude()))


def alignElementInZeroPoint(elements_in,ref_element):
    
        
    xg = cw.point_3d(1,0,0)
    yg = cw.point_3d(0,1,0)
    zg = cw.point_3d(0,0,1)
    zero_point = cw.point_3d(0,0,0)
    
    
    p1 = gc.get_p1(ref_element)
    ec.move_element(elements_in, -p1)
    
    # Get the X-vector of the reference element    
    xl = gc.get_xl(ref_element)

    # Rotate elements ato align the global X-axis
    ec.rotate_elements(elements_in, zero_point, xl.cross(xg), angle_between_vectors(xl, xg))
            
    # Rotate elements ato align the global Z-axis
    zl = gc.get_zl(ref_element)
    ec.rotate_elements(elements_in, zero_point, zl.cross(zg), angle_between_vectors(zl, zg))

    # Move the reference object above 0 in Y and Z direction
    ec.move_element(elements_in, xg * -gc.get_length(ref_element)/2)
    #ec.move_element([element_in], zg * gc.get_height(element_in)/2)






# get active element_ids
element_ids = ec.get_active_identifiable_element_ids()


# EDIT HERE BY WHAT FUNCTION YOU WANT TO GROUP
grouped_elements = group_elements_by_function(element_ids,ac.get_subgroup)


dir_Y = cw.point_3d(0,1,0)
dir_Z = cw.point_3d(0,0,1)

last_pos = cw.point_3d(0,0,0)

for key in grouped_elements.keys():
    print(key)
    grouped_elements[key] = sort_elements_by_function(grouped_elements[key],gc.get_volume)

    elements_in_drawing = grouped_elements[key]

    ref_element = elements_in_drawing[-1]
    
    alignElementInZeroPoint(elements_in_drawing,ref_element)

    ec.move_element(elements_in_drawing,dir_Z*gc.get_height(ref_element)/2 + last_pos)

    last_pos += dir_Y* 1000

    if len(elements_in_drawing) >1:
        ec.join_elements(elements_in_drawing)
```

**Reply by roehligas** [reply]
In your case i guess you can use your own dictionary that you already have... :)

**Reply by lucas-ita** [reply]
That was helpful. It's not quite what I need yet, but it's a good start. Thank you very much!

**Reply by Brunner246** [reply]
I guess the easiest thing to do is a `copy` / `move` and `rotate` as suggested by @roehligas 

Alternatively, you can cache the translation vectors of e.g. point 1 to the point 1 of the reference element and the axis orientation based on the reference element and then transform the component using these values.
